File: DebugIntro/C/03_Overflows.c
#include <stdio.h>
#include <float.h>
#include <time.h>
#include <limits.h>
#include <float.h>

int main(int argc, char** argv){

  int i = 0;
  int j = INT_MAX, k = INT_MIN;
  int l = UINT_MAX;

  float a = FLT_MIN, b = FLT_MAX, bb;
  double c = DBL_MIN, d = DBL_MAX;

  printf("A normal 4 byte integer must be between INT_MIN of %d and INT_MAX of %d\n", k, j);
  printf("Adding one to INT_MAX: %d + 1 = %d\n", j, j+1);
  printf("Subtracting one from INT_MIN: %d - 1 = %d\n", k, k-1);

  printf("\n");

  printf("C also has unsigned integers which are between %d and %u\n", i, l);
  printf("Adding one to UINT_MAX is undefined behaviour!\n");
  printf("Here it is: %u + 1 = %u \n", l, l+1);
  printf("\n");
  printf("Bonus points: what is this line doing? %u = %d\n", l, l);

  printf("\n");

  printf("Floating point numbers also have limits.\n");
  printf("The min and max float are %10.6e and %10.6e\n", a, b);

  bb = b * 1.1;
  printf("FLT_MAX * 1.1 is %10.6e\n", bb);

  printf("Bonus: the following is unexpected! See first note.");
  bb = b + 1.0;
  printf("FLT_MAX + 1 is %10.6e\n", bb);


  printf("Bonus: the following is unexpected! See second note.");
  bb = a /1.1;
  printf("FLT_MIN / 1.1 is %10.6e\n", bb);


}

File: DebugIntro/Fortran/03_Overflows.f90

MODULE KINDS

  !Constants for "single" and "double" precision floats
  INTEGER, PARAMETER :: real32 = KIND(1.0)
  INTEGER, PARAMETER :: real64 = KIND(1.0d0)

END MODULE KINDS


PROGRAM MAIN

  USE KINDS

  INTEGER :: i = 0
  INTEGER :: j = HUGE(1), k = -HUGE(1) - 1
  REAL(KIND=real32) :: a = TINY(1.0_real32), b = HUGE(1.0_real32), bb
  REAL(KIND=real64) :: c = TINY(1.0_real64), d = HUGE(1.0_real64)

  WRITE(*, fmt='(a41, i0, a10, i0)') "A normal 4 byte integer must be between ", &
      k, "and HUGE ", j
  WRITE(*, fmt='(a21, i0, a7, i0)') "Adding one to HUGE: ", j,"+ 1 = ", j + 1
  WRITE(*, fmt='(a33, i0, a7, i0)') "Subtracting one from -(HUGE+1): ", &
      k,"- 1 = ", k - 1

  PRINT*, ''
  PRINT*, "Floating point numbers also have limits."

  WRITE(*, fmt='(a27, e12.6, a5, e12.6)') "The min and max float are ", a, " and ", b


  bb = b * 1.1
  WRITE(*, fmt='(a15, e12.6)') "HUGE * 1.1 is ", bb

  bb = b  + 1
  PRINT*, "Bonus: the following is unexpected! See first note."
  WRITE(*, fmt='(a15, e12.6)') "HUGE + 1.1 is ", bb


  PRINT*, "Bonus: the following is unexpected! See second note."
  bb = a /1.1
  WRITE(*, fmt='(a15, e12.6)') "TINY / 1.1 is ", bb


END PROGRAM MAIN
